Prevent Klipper from storing secrets in clipboard history

TheZ3ro opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Description
Fix #584 
Porting of https://github.com/keepassx/keepassx/pull/211 "Add hint for Klipper to not add passwords to history" by @roberthoffmann
Relevant Klipper issue/PR https://phabricator.kde.org/D12539

<!--- Describe your changes in detail -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Klipper adds every item to the clipboard history, unless an additional mime type `'x-kde-passwordManagerHint` is set to `secret`. 

## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I cannot test this since I don't have klipper installed.
Maybe @phoerious or @Germano0 can test this

## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Please remove all lines which don't apply. -->
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
<!--- Please go over all the following points. -->
<!--- Again, remove any lines which don't apply. -->
<!--- Pull Requests that don't fulfill all [REQUIRED] requisites are likely -->
<!--- to be sent back to you for correction or will be rejected.  -->
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

phoerious commented on src/gui/Clipboard.h:
```diff
@@ -31,7 +31,8 @@ class Clipboard : public QObject
 
 public:
-    void setText(const QString& text);
+    void setText(const QString& text, bool secret = false);
```
You also need to replace all usages of setText() with the call to the secret version.

TheZ3ro replied:
Why? Non secret field will be visible. Only passwords will be omitted from the clipboard history. 
I've already changed the copy-password to clipboard usages to use the secret version. 
Other fields will call setText with secret set to false by default

phoerious replied:
I don't see any use of the new functions in this PR, except in setClipboardSecretTextAndMinimize() (which is also new).
Besides, I would prefer this flag to be set for everything. Usernames etc. are also cleared from the clipboard after the timeout, so they should also not be in the history.

TheZ3ro replied:
setClipboardSecretTextAndMinimize() itself it's used when copying the password

I will change it since I agree that all field should be treated like secrets

Germano0 replied:
If you have any dubts, you may want to ask Martin Flöser, the Klipper developer.
When the patch will be finished, please let me know, so I will test it

phoerious commented:
Somehow this PR shrunk. What happened to the rest of the changes and to what I nagged about in my first review?

TheZ3ro commented:
@phoerious look at the difference between the two commits.

In the old commit:
- I defined a new `setSecretText` function in the `Clipboard` class with the secret MIMEtype (used by Klipper to delete entries from history)
- I defined a new `setClipboardSecretTextAndMinimize` function in the `DatabaseWidget` class calling the `setSecretText` one
- I used the `setClipboardSecretTextAndMinimize` function whenever the password was copied to clipboard

In the new commit:
- I edited the already existing `setText` function in the `Clipboard` class to always expose the secret MIMEtype when copying any text to clipboard. 
Klipper will catch this and delete the entries from his history but other clipboard manager will happily ignore and perform as standard

IMHO it seems pretty easy to understand from the code

phoerious commented:
All I see in the Diff are +11 -3

TheZ3ro commented:
Checkout the `feature/klipper-password` branch, do a `git log` to see the 2 commit i'm talking about:

> commit a4f79232ff0a9266416bfdbd82670ec89e32aa06
> Author: thez3ro
> Date:   Tue May 29 22:35:00 2018 +0200
>    all fields are protected
> 
> commit 9f244ebb12e3ce8e6bd86b20d3e2a590da136589
> Author: thez3ro
> Date:   Wed May 16 15:00:41 2018 +0200
>    prevent klipper from storing password in clipboard history


Then you can:
1) use `git diff HEAD~2 HEAD~1` to display the second-last commit (labeled old in my previous comment)
2) use `git diff HEAD~2 HEAD` to display the final branch status +11 -3 (labeled new in my previous comment)
3) use `git diff HEAD~1 HEAD` to display only the last commit where I revert changes made in **1**

On the next rebase, I will squash those 2 commit into 1 so the revert part will not be present in git history

-----

Alternatively in the commit tab you can see the two specific commit 
![istantanea_2018-06-12_00-24-53](https://user-images.githubusercontent.com/686894/41260273-162235a4-6dd7-11e8-9b97-627c78f1dfee.png)

TheZ3ro commented:
Can someone review this? @keepassxreboot/core-developers

droidmonkey commented on src/gui/Clipboard.cpp:
```diff
@@ -46,15 +48,20 @@ void Clipboard::setText(const QString& text)
-    clipboard->setText(text, QClipboard::Clipboard);
+    const QString secretStr = "secret";
+    QByteArray secretBa = secretStr.toUtf8();
+    mime->setText(text);
```
Can we move this line and line 53 above the #ifdef, they are the same regardless of OSX or not

TheZ3ro commented:
@phoerious What?

I've deleted the older branch, created a new one with the same name, added the new changes, did a rebase excluding the old commits and force-push.
This is a normal thing in a git environment and it has been done many times before on this repo, nothing strange.

It's useless to keep in a branch waiting for a merge old changes that are being reverted the next commit.
It only use extra repository space in the commit history and in git objects.

So, what are you talking about?

Checkout the branch with your git client, what you see is what gets merged.

shantanugoel commented:
Is there any more work to be done on this PR before this can be merged? Looks like @droidmonkey approved the changes and the change requested by @phoerious is not applicable any more?

droidmonkey commented:
It looks complete to me, I need to test it using klipper before I merge unless someone can get certainty that it works.

shantanugoel commented:
Thanks @droidmonkey. I built the develop branch just now along with this pull request and tested on ArchLinux. I can confirm that klipper now ignores the password copied from keepassxc.

sts10 commented:
Is there any chance that this fix doesn't work for Snap installations? 

I ask because, unfortunately, running 2.4.0 installed as a Snap on Kubuntu 18.04, my passwords remain in my clipboard  indefinitely (rather than 10 seconds, as expected). 

I'm more than happy to file a fresh issue for this, but just wanted to check if (a) there was something I had to do in KDE preferences to allow KPXC to clear the clipboard or (b) this wasn't tested on a Snap install and that could be the issue.

shantanugoel commented:
@sts10 At least I tested it on a regular install only, not on snap. And it's working with the official 2.4.0 release as well. Are you sure you have the right version of klipper installed that has the counterpart patch to look at the metadata sent by keepassxc and ignore it? That patch (https://phabricator.kde.org/D12539) was merged only around May last year but you seem to be on Kubuntu 18.04, so there's a high chance that it's not present in your OS.

kozaq commented:
Hi, tested KeePassXC 2.4.0 and 2.4.1 on Kubuntu 18.04.2 LTS with password set to be cleared from clipboard after 10 seconds. Sadly this does not work. What I see is the lack of Klipper - instead there is Plasma's own clipboard app. 

**OS:** 
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.2 LTS"
VERSION_ID="18.04"

**Kernel release:** 
4.15.0-47-generic

**Plasma version: **
5.12.7

Is there a chance that Plasma's clipboard will be supported?

droidmonkey commented:
Open a new issue if you want to see that support.

roberthoffmann commented:
From plasma-workspace 5.12.8 the "x-kde-passwordManagerHint" is supported, so in Ubuntu Cosmic (which has 5.13.5) it should work.
